alembic/versions/68ce6a306b30_add_koji_build_groups.py

```diff
@@ -0,0 +1,443 @@
+    status = Column(Enum(ProposeDownstreamStatus))
+    submitted_time = Column(DateTime, default=datetime.utcnow)
+
+    runs = relationship("PipelineModel", back_populates="propose_downstream_run")
+    propose_downstream_targets = relationship(
+        "ProposeDownstreamTargetModel", back_populates="propose_downstream"
+    )
+
+
+class GroupModel:
+    @property
+    def grouped_targets(self):
+        raise NotImplementedError
+
+
+class TFTTestRunGroupModel(ProjectAndTriggersConnector, GroupModel, Base):
+    __tablename__ = "tft_test_run_groups"
+    id = Column(Integer, primary_key=True)
+    submitted_time = Column(DateTime, default=datetime.utcnow)
+
+    runs = relationship("PipelineModel", back_populates="test_run_group")
+    tft_test_run_targets = relationship(
+        "TFTTestRunTargetModel", back_populates="group_of_targets"
+    )
+
+    @property
+    def grouped_targets(self) -> List["TFTTestRunTargetModel"]:
+        return self.tft_test_run_targets
+
+
+class CoprBuildGroupModel(ProjectAndTriggersConnector, GroupModel, Base):
+    __tablename__ = "copr_build_groups"
+    id = Column(Integer, primary_key=True)
+    submitted_time = Column(DateTime, default=datetime.utcnow)
+
+    runs = relationship("PipelineModel", back_populates="copr_build_group")
+    copr_build_targets = relationship(
+        "CoprBuildTargetModel", back_populates="group_of_targets"
+    )
+
+    @property
+    def grouped_targets(self) -> List["CoprBuildTargetModel"]:
+        return self.copr_build_targets
+
+
+class KojiBuildGroupModel(ProjectAndTriggersConnector, GroupModel, Base):
+    __tablename__ = "koji_build_groups"
+    id = Column(Integer, primary_key=True)
+    submitted_time = Column(DateTime, default=datetime.utcnow)
+
+    runs = relationship("PipelineModel", back_populates="koji_build_group")
+    koji_build_targets = relationship(
+        "KojiBuildTargetModel", back_populates="group_of_targets"
+    )
+
+    @property
+    def grouped_targets(self):
+        return self.koji_build_targets
+
+
+def upgrade():
+    bind = op.get_bind()
+    session = sa.orm.Session(bind=bind)
+
+    op.create_table(
+        "koji_build_groups",
+        sa.Column("id", sa.Integer(), primary_key=True),
+        sa.Column("submitted_time", sa.DateTime(), nullable=True),
+    )
+    op.add_column(
+        "koji_build_targets",
+        sa.Column("koji_build_group_id", sa.Integer(), nullable=True),
+    )
+    op.create_foreign_key(
+        None,
+        "koji_build_targets",
+        "koji_build_groups",
+        ["koji_build_group_id"],
+        ["id"],
+    )
+    op.add_column(
+        "pipelines", sa.Column("koji_build_group_id", sa.Integer(), nullable=True)
+    )
+    op.create_foreign_key(
+        None, "pipelines", "koji_build_groups", ["koji_build_group_id"], ["id"]
+    )
+
+    # Group by the same SRPM
+    srpm_ids = collections.defaultdict(list)
+    for koji_build in session.query(KojiBuildTargetModel).all():
```

Comment:
The `Query` is already iterable so you don't need `all()` (to create a list from it first) if you are going to just iterate over it, see [for example](https://github.com/packit/packit-service/blob/6f648528db0535b9bf798b814efb74a5fc895139/packit_service/models.py#L1210)

```suggestion
    for koji_build in session.query(KojiBuildTargetModel):
```